File: ios/Ubapp/Games/Tag/TagEngine.swift
import Foundation

enum PlayerStatus { case runner, it, frozen, eliminated }

final class TagPlayerView {
    let id: String
    let displayName: String
    var status: PlayerStatus
    init(id: String, displayName: String, status: PlayerStatus) {
        self.id = id; self.displayName = displayName; self.status = status
    }
}

final class TagState {
    let variant: TagVariant
    var players: [String: TagPlayerView]
    let startedAtMs: Int64
    let deadlineMs: Int64
    var endReason: String?
    var winnerId: String?

    init(variant: TagVariant, players: [String: TagPlayerView], startedAtMs: Int64, deadlineMs: Int64) {
        self.variant = variant; self.players = players
        self.startedAtMs = startedAtMs; self.deadlineMs = deadlineMs
    }

    var isOver: Bool { endReason != nil }
    var its: [TagPlayerView] { players.values.filter { $0.status == .it } }
    var runners: [TagPlayerView] { players.values.filter { $0.status == .runner } }
    var frozen: [TagPlayerView] { players.values.filter { $0.status == .frozen } }
    var alive: [TagPlayerView] { players.values.filter { $0.status != .eliminated } }
}

/// Deterministic state machine — given the same `start` + ordered events,
/// every device computes the same state.
final class TagEngine {
    let selfId: String
    var state: TagState?

    init(selfId: String) { self.selfId = selfId }

    @discardableResult
    func start(variant: TagVariant, startingItId: String, startTimeMs: Int64,
               peerIds: [String], displayNames: [String: String]) -> TagState {
        var players: [String: TagPlayerView] = [:]
        for id in peerIds {
            let name = displayNames[id] ?? String(id.prefix(6))
            players[id] = TagPlayerView(
                id: id, displayName: name,
                status: id == startingItId ? .it : .runner)
        }
        let durationMs: Int64 = variant == .hotPotato
            ? 10 * 60 * 1000
            : Int64(variant.duration * 1000)
        let s = TagState(variant: variant, players: players,
                         startedAtMs: startTimeMs, deadlineMs: startTimeMs + durationMs)
        state = s
        return s
    }

    @discardableResult
    func applyTag(taggerId: String, victimId: String) -> Bool {
        guard let s = state, !s.isOver,
              let tagger = s.players[taggerId], let victim = s.players[victimId],
              tagger.status == .it, victim.status == .runner else { return false }
        switch s.variant {
        case .classic, .bomb, .hotPotato:
            tagger.status = .runner
            victim.status = .it
        case .freeze:
            victim.status = .frozen
            if s.runners.isEmpty { s.endReason = "all_frozen"; s.winnerId = tagger.id }
        case .zombie:
            victim.status = .it
            if s.runners.isEmpty { s.endReason = "last_survivor" }
        }
        return true
    }

    @discardableResult
    func applyUnfreeze(unfreezerId: String, victimId: String) -> Bool {
        guard let s = state, !s.isOver, s.variant == .freeze,
              let u = s.players[unfreezerId], let v = s.players[victimId],
              u.status == .runner, v.status == .frozen else { return false }
        v.status = .runner
        return true
    }

    func applyEnd(reason: String, winnerId: String?) {
        guard let s = state, !s.isOver else { return }
        s.endReason = reason
        s.winnerId = winnerId
    }

    /// Hot-potato only: if this device is "it" when the per-tag countdown expires.
    func hotPotatoTimeout() -> (reason: String, winnerId: String?)? {
        guard let s = state, !s.isOver, s.variant == .hotPotato,
              let me = s.players[selfId], me.status == .it else { return nil }
        me.status = .eliminated
        return ("hot_potato_timeout", s.alive.first?.id)
    }
}

File: android/app/src/main/java/com/example/ubapp/games/tag/TagEngine.kt
package com.example.ubapp.games.tag

enum class PlayerStatus { RUNNER, IT, FROZEN, ELIMINATED }

class TagPlayerView(val id: String, val displayName: String, var status: PlayerStatus)

class TagState(
    val variant: TagVariant,
    val players: MutableMap<String, TagPlayerView>,
    val startedAtMs: Long,
    val deadlineMs: Long,
) {
    var endReason: String? = null
    var winnerId: String? = null
    val isOver: Boolean get() = endReason != null
    val its get() = players.values.filter { it.status == PlayerStatus.IT }
    val runners get() = players.values.filter { it.status == PlayerStatus.RUNNER }
    val frozen get() = players.values.filter { it.status == PlayerStatus.FROZEN }
    val alive get() = players.values.filter { it.status != PlayerStatus.ELIMINATED }
}

/**
 * Deterministic state machine — given the same `start` + ordered events,
 * every device computes the same state.
 */
class TagEngine(val selfId: String) {
    var state: TagState? = null

    fun start(
        variant: TagVariant, startingItId: String, startTimeMs: Long,
        peerIds: List<String>, displayNames: Map<String, String>,
    ): TagState {
        val players = LinkedHashMap<String, TagPlayerView>()
        for (id in peerIds) {
            val name = displayNames[id] ?: id.take(6)
            players[id] = TagPlayerView(
                id, name,
                if (id == startingItId) PlayerStatus.IT else PlayerStatus.RUNNER,
            )
        }
        val duration = if (variant == TagVariant.HOT_POTATO) 10 * 60_000L else variant.durationMs
        val s = TagState(variant, players, startTimeMs, startTimeMs + duration)
        state = s
        return s
    }

    fun applyTag(taggerId: String, victimId: String): Boolean {
        val s = state ?: return false
        if (s.isOver) return false
        val tagger = s.players[taggerId] ?: return false
        val victim = s.players[victimId] ?: return false
        if (tagger.status != PlayerStatus.IT || victim.status != PlayerStatus.RUNNER) return false
        when (s.variant) {
            TagVariant.CLASSIC, TagVariant.BOMB, TagVariant.HOT_POTATO -> {
                tagger.status = PlayerStatus.RUNNER; victim.status = PlayerStatus.IT
            }
            TagVariant.FREEZE -> {
                victim.status = PlayerStatus.FROZEN
                if (s.runners.isEmpty()) { s.endReason = "all_frozen"; s.winnerId = tagger.id }
            }
            TagVariant.ZOMBIE -> {
                victim.status = PlayerStatus.IT
                if (s.runners.isEmpty()) s.endReason = "last_survivor"
            }
        }
        return true
    }

    fun applyUnfreeze(unfreezerId: String, victimId: String): Boolean {
        val s = state ?: return false
        if (s.isOver || s.variant != TagVariant.FREEZE) return false
        val u = s.players[unfreezerId] ?: return false
        val v = s.players[victimId] ?: return false
        if (u.status != PlayerStatus.RUNNER || v.status != PlayerStatus.FROZEN) return false
        v.status = PlayerStatus.RUNNER
        return true
    }

    fun applyEnd(reason: String, winnerId: String?) {
        val s = state ?: return
        if (s.isOver) return
        s.endReason = reason; s.winnerId = winnerId
    }

    fun hotPotatoTimeout(): Pair<String, String?>? {
        val s = state ?: return null
        if (s.isOver || s.variant != TagVariant.HOT_POTATO) return null
        val me = s.players[selfId] ?: return null
        if (me.status != PlayerStatus.IT) return null
        me.status = PlayerStatus.ELIMINATED
        return "hot_potato_timeout" to s.alive.firstOrNull()?.id
    }
}
